rewrite-java/src/main/java/org/openrewrite/java/JavaTemplate.java

```diff
@@ -45,13 +43,16 @@
     private final int parameterCount;
     private final Set<String> imports;
     private final String parameterMarker;
+    @Nullable
+    private final StringWriter tap;
 
-    private JavaTemplate(JavaParser parser, String code, Set<String> imports, String parameterMarker) {
+    private JavaTemplate(JavaParser parser, String code, Set<String> imports, String parameterMarker, @Nullable StringWriter tap) {
```

Comment:
`Consumer<String>` on `tap` is more flexible? Maybe be more explicit then `Consumer<String> onBeforeParseTemplate`. Or even more descriptive:

```java
interface TemplateEventHandler {
   // so they can see just their template with variables substituted
   void afterVariableSubstitution(String substitutedTemplate);

   // so they can see their substituted template inserted into its surrounding context
   void beforeParseTemplate(String generatedTemplated)
}
```